hw/mcu/nordic/nrf52xxx/src/hal_qspi.c

```diff
@@ -0,0 +1,279 @@
+static int
+nrf52k_qspi_init(const struct hal_flash *dev);
+
+static const struct hal_flash_funcs nrf52k_qspi_funcs = {
+    .hff_read = nrf52k_qspi_read,
+    .hff_write = nrf52k_qspi_write,
+    .hff_erase_sector = nrf52k_qspi_erase_sector,
+    .hff_sector_info = nrf52k_qspi_sector_info,
+    .hff_init = nrf52k_qspi_init
+};
+
+const struct hal_flash nrf52k_qspi_dev = {
+    .hf_itf = &nrf52k_qspi_funcs,
+    .hf_base_addr = 0x00000000,
+    .hf_size = MYNEWT_VAL(QSPI_FLASH_SECTOR_COUNT) * MYNEWT_VAL(QSPI_FLASH_SECTOR_SIZE),
+    .hf_sector_cnt = MYNEWT_VAL(QSPI_FLASH_SECTOR_COUNT),
+    .hf_align = 1
+};
+
+static int
+nrf52k_qspi_read(const struct hal_flash *dev, uint32_t address,
+        void *dst, uint32_t num_bytes) {
+    uint32_t ram_buffer[4];
+    uint8_t *ram_ptr = NULL;
+    uint32_t read_bytes;
+
+    while ((NRF_QSPI->STATUS & QSPI_STATUS_READY_Msk) == 0)
+        ;
+
+    while (num_bytes != 0) {
+        /*
+         * If address or destination pointer are unaligned or
+         * number of bytes to read is less then 4 read using ram buffer.
+         */
+        if (((address & 3) != 0) ||
+            ((((uint32_t) dst) & 3) != 0) ||
+            num_bytes < 4) {
+            uint32_t to_read = (num_bytes + (address & 3) + 3) & ~3;
+            if (to_read > sizeof(ram_buffer)) {
+                to_read = sizeof(ram_buffer);
+            }
+            ram_ptr = (uint8_t *)((uint32_t) ram_buffer + (address & 3));
+            read_bytes = to_read - (address & 3);
+            if (read_bytes > num_bytes) {
+                read_bytes = num_bytes;
+            }
+            NRF_QSPI->READ.DST = (uint32_t) ram_buffer;
+            NRF_QSPI->READ.SRC = address & ~3;
+            NRF_QSPI->READ.CNT = to_read;
+        } else {
+            read_bytes = num_bytes & ~3;
+            NRF_QSPI->READ.DST = (uint32_t) dst;
+            NRF_QSPI->READ.SRC = address;
+            NRF_QSPI->READ.CNT = read_bytes;
+        }
+        NRF_QSPI->EVENTS_READY = 0;
+        NRF_QSPI->TASKS_READSTART = 1;
+        while (NRF_QSPI->EVENTS_READY == 0)
+            ;
+        if (ram_ptr != NULL) {
+            memcpy(dst, ram_ptr, read_bytes);
+            ram_ptr = NULL;
+        }
+        address += read_bytes;
+        dst = (void *) ((uint32_t) dst + read_bytes);
+        num_bytes -= read_bytes;
+    }
+    return 0;
+}
+
+static int
+nrf52k_qspi_write(const struct hal_flash *dev, uint32_t address,
+        const void *src, uint32_t num_bytes)
+{
+    uint32_t ram_buffer[4];
+    uint8_t *ram_ptr = NULL;
+    uint32_t written_bytes;
+    const char src_not_in_ram = (((uint32_t) src) & 0xE0000000) != 0x20000000;
+
+    while ((NRF_QSPI->STATUS & QSPI_STATUS_READY_Msk) == 0)
+        ;
+
+    while (num_bytes != 0) {
+        const uint32_t page_start = address
+                & ~(MYNEWT_VAL(QSPI_FLASH_PAGE_SIZE) - 1);
+        const uint32_t page_limit = page_start
+                + MYNEWT_VAL(QSPI_FLASH_PAGE_SIZE);
```

Comment:
Coding style: move the page_start, page_limit declaration in the beginning of the function.
When you have a statement which expands to multiple lines, put the operator at the end of the line.
See coding style doc in the root of the repository.